File: src/meta_schedule/measure_callback.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./measure_callback.h"  // NOLINT(build/include)

#include <dmlc/memory_io.h>
#include <tvm/node/serialization.h>
#include <tvm/runtime/registry.h>

#include <fstream>

#include "../support/base64.h"

namespace tvm {
namespace meta_schedule {

/********** RecordToFile **********/

RecordToFile::RecordToFile() { data_ = make_object<RecordToFileNode>(); }

void RecordToFileNode::Init(const SearchTask& task) {
  ICHECK(!task->log_file.value_or("").empty())
      << "ValueError: log_file is not specified in SeachTask";
  this->log_file = task->log_file.value();
  this->task_name = task->task_name;
  this->target = task->target->Export();
  this->target_host = task->target_host->Export();
  {
    std::string prim_func_json = SaveJSON(task->workload);
    std::string prim_func_b64;
    dmlc::MemoryStringStream m_stream(&prim_func_b64);
    support::Base64OutStream b64strm(&m_stream);
    dmlc::Stream* strm = &b64strm;
    strm->Write(prim_func_json);
    b64strm.Finish();
    this->prim_func_b64 = prim_func_b64;
  }
}

void RecordToFileNode::Callback(const Array<MeasureInput>& inputs,
                                const Array<MeasureResult>& results) {
  static const auto* f_serialize = runtime::Registry::Get("meta_schedule._serialize_json");
  ICHECK(f_serialize) << "IndexError: Cannot find packed function \""
                         "meta_schedule._serialize_json\", which should be registered in python";
  CHECK(!this->log_file.empty()) << "ValueError: empty log_file for measure logs";
  std::ofstream ofs(this->log_file, std::ofstream::app);
  ICHECK_EQ(inputs.size(), results.size());
  for (int i = 0, n = inputs.size(); i < n; ++i) {
    const MeasureInput& measure_input = inputs[i];
    const MeasureResult& measure_result = results[i];
    MeasureErrorNO error_no = static_cast<MeasureErrorNO>(measure_result->error_no);
    if (error_no != MeasureErrorNO::kNoError) {
      continue;
    }
    Array<ObjectRef> result{
        this->task_name,                                    // record[0]
        this->target,                                       // record[1]
        this->target_host,                                  // record[2]
        measure_result->costs,                              // record[3]
        measure_input->sch->trace().value()->AsJSON(true),  // record[4]
        String(kLogVersion),                                // record[5]
        this->prim_func_b64,                                // record[6]
    };
    String record = (*f_serialize)(result);
    ofs << record << std::endl;
  }
}

/********** FFI **********/

struct Internal {
  static RecordToFile RecordToFileNew() { return RecordToFile(); }
};

TVM_REGISTER_NODE_TYPE(RecordToFileNode);
TVM_REGISTER_GLOBAL("meta_schedule.RecordToFile").set_body_typed(Internal::RecordToFileNew);

}  // namespace meta_schedule
}  // namespace tvm

File: python/tvm/meta_schedule/measure_callback/measure_callback.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Meta Schedule MeasureCallback."""

from typing import TYPE_CHECKING, List

from tvm._ffi import register_object
from tvm.runtime import Object
from tvm.meta_schedule import TuneContext
from tvm.meta_schedule.search_strategy import MeasureCandidate
from tvm.meta_schedule.builder import BuilderResult
from tvm.meta_schedule.runner import RunnerResult
from tvm.meta_schedule.utils import _get_hex_address

from .. import _ffi_api

if TYPE_CHECKING:
    from ..tune_context import TuneContext
    from ..task_scheduler import TaskScheduler


@register_object("meta_schedule.MeasureCallback")
class MeasureCallback(Object):
    """Rules to apply after measure results is available."""

    def apply(
        self,
        task_scheduler: "TaskScheduler",
        tasks: List["TuneContext"],
        measure_candidates: List[MeasureCandidate],
        builds: List[BuilderResult],
        results: List[RunnerResult],
    ) -> bool:
        """Apply a measure callback to the given schedule.

        Parameters
        ----------
        task_scheduler: TaskScheduler
            The task scheduler.
        tasks: List[TuneContext]
            The list of tune context to process.
        measure_candidats: List[MeasureCandidate]
            The measure candidates.
        builds: List[BuilderResult]
            The builder results by building the measure candidates.
        results: List[RunnerResult]
            The runner results by running the built measure candidates.

        Returns
        -------
        result : bool
            Whether the measure callback was successfully applied.
        """
        return _ffi_api.MeasureCallbackApply(
            self, task_scheduler, tasks, measure_candidates, builds, results
        )


@register_object("meta_schedule.PyMeasureCallback")
class PyMeasureCallback(MeasureCallback):
    """An abstract MeasureCallback with customized methods on the python-side."""

    def __init__(self):
        """Constructor."""

        def f_apply(
            task_scheduler: "TaskScheduler",
            tasks: List[TuneContext],
            measure_candidates: List[MeasureCandidate],
            builds: List[BuilderResult],
            results: List[RunnerResult],
        ) -> bool:
            return self.apply(task_scheduler, tasks, measure_candidates, builds, results)

        def f_as_string() -> str:
            return str(self)

        self.__init_handle_by_constructor__(
            _ffi_api.MeasureCallbackPyMeasureCallback,  # type: ignore # pylint: disable=no-member
            f_apply,
            f_as_string,
        )

    def __str__(self) -> str:
        return f"PyMeasureCallback({_get_hex_address(self.handle)})"
